content/advent-2017/tracking-the-stars.md

```diff
@@ -0,0 +1,286 @@
+on Amazon Web Services using a timer that triggers an execution every 15 minutes. The lambda
+function collects the stars count of the top 1000 Go repositories, which is the maximum the GitHub
+search API will return. This is a post about Go, so the lambda function is written in... Python, of
+course! The total resource usage is well below the free tier limits, so running this function is
+free indefinitely.
+
+The data that the lambda function collects is stored in a DynamoDB table that just holds the repo
+name, an integer timestamp (epoch time in seconds), and the number of stars at that time. By the
+time this article was started, I had several hundred megabytes of data in DynamoDB. As with the
+Lambda function, the usage here is below the free tier limit, so the storage is free as well. For
+reference, the allocated read and write capacity is 20 units each.
+
+Just getting the data out of DynamoDB was a chore, as I chose to use the Data Pipeline in AWS which
+runs a MapReduce cluster on your behalf to extract data from DynamoDB into S3. There is a standard
+template to dump a DynamoDB table into S3 that's fairly easy to figure out. From there, I had to
+download it all locally using the AWS CLI and then write a program (this time in Go) to convert the
+many separate chunk files into one CSV file. This left me with a 726 MB CSV file.
+
+The [go program](/postimages/advent-2017/tracking-the-stars/tocsv.go) to do the CSV conversion is
+incredibly over-engineered, but it was fun to optimize while the next step completed.
+
+I decided that a real database was going to be helpful in doing the analysis, so I set up Postgres
+locally and imported the giant CSV file into a table. The following is the schema and command to
+copy the CSV data into Postgres. This would be run at the `psql` command line.
+
+```sql
+drop table if exists github_stars;
+
+create unlogged table github_stars (
+    repo  varchar(255) not null,
+    ts    integer      not null,
+    stars integer      not null,
+    constraint repo_ts primary key(repo,ts)
+);
+
+create index repo_idx  on github_stars(repo);
+create index ts_idx    on github_stars(ts);
+create index stars_idx on github_stars(stars);
+
+\copy github_stars from 'github-stars.csv' csv
+```
+
+## Let's dig in
+
+At the end of this (lengthy) import I was left with a relatively simple table that has 23,279,479
+rows of data about 1412 separate repositories. Note that this number isn't easily divided by 1000.
+That's because the error handling was not extremely robust in the lambda. It's designed for long
+term trends, not second-by-second updates. The number of tracked repositories is higher than 1000
+because some repositories stayed the same and slipped out of view as others increased. Yet others
+increased rapidly from 0 to attain a relatively high position in the middle of tracking.
+
+I should warn anyone that is halfway decent at SQL that my skills are very rusty and therefore you
```

Comment:
I think you can probably omit this sentence.  Anyone who's interested in criticizing your SQL is missing the point of the blog. :)

Reply:
You're right. You're the third person to say that, nixed.